File: ide/vscode/ipc/endpoint.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

const ENDPOINT_FILE = "ide-ipc.json";

/**
 * Local Agent가 state dir에 쓰는 IDE IPC 디스커버리 정보.
 * 워크스페이스 파일이 아니며, 이 모듈은 이 파일만 읽는다.
 */
export interface IpcEndpoint {
  bind: string;
  port: number;
  path: string;
  token: string;
}

/**
 * ide-ipc.json 본문을 검증한다. bind/path/token이 비었거나 port가 1 미만이면 거부한다.
 *
 * @param raw UTF-8 JSON 문자열. 빈 값은 안 된다.
 * @returns 파싱된 endpoint
 */
export function parseEndpoint(raw: string): IpcEndpoint {
  const parsed = JSON.parse(raw) as Partial<IpcEndpoint>;
  if (
    typeof parsed.bind !== "string" ||
    parsed.bind.trim() === "" ||
    typeof parsed.port !== "number" ||
    parsed.port < 1 ||
    typeof parsed.path !== "string" ||
    parsed.path.trim() === "" ||
    typeof parsed.token !== "string" ||
    parsed.token.trim() === ""
  ) {
    throw new Error("incomplete ide ipc endpoint");
  }
  return {
    bind: parsed.bind,
    port: parsed.port,
    path: parsed.path,
    token: parsed.token,
  };
}

/**
 * Local Agent state dir을 고른다. 워크스페이스 루트가 아니라 에이전트 설정 디렉터리다.
 *
 * @param override 확장 설정값. 빈 문자열이면 환경변수/플랫폼 기본값
 */
export function resolveStateDir(override?: string): string {
  if (override && override.trim() !== "") {
    return path.resolve(override);
  }
  if (process.env.ONCODE_AGENT_STATE_DIR) {
    return path.resolve(process.env.ONCODE_AGENT_STATE_DIR);
  }
  return path.join(userConfigDir(), "oncode-agent");
}

/**
 * state dir의 ide-ipc.json을 동기 읽기한다. 없거나 깨졌으면 null이다.
 * 워크스페이스 소스는 읽지 않는다.
 *
 * @param stateDir Local Agent state dir 절대 경로
 */
export function readEndpointFile(stateDir: string): IpcEndpoint | null {
  const file = path.join(stateDir, ENDPOINT_FILE);
  try {
    const raw = fs.readFileSync(file, "utf8");
    return parseEndpoint(raw);
  } catch {
    return null;
  }
}

/** Go os.UserConfigDir()과 같은 플랫폼 설정 루트. */
function userConfigDir(): string {
  if (process.platform === "win32") {
    return process.env.APPDATA || path.join(os.homedir(), "AppData", "Roaming");
  }
  if (process.platform === "darwin") {
    return path.join(os.homedir(), "Library", "Application Support");
  }
  return process.env.XDG_CONFIG_HOME || path.join(os.homedir(), ".config");
}

File: local-agent/internal/idebridge/endpoint.go
package idebridge

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
)

// EndpointFileName은 Local Agent state dir에 쓰는 IDE IPC 디스커버리 파일이다.
const EndpointFileName = "ide-ipc.json"

// Endpoint는 확장이 WebSocket에 붙는 데 필요한 연결 정보다.
// Token은 에이전트 재시작마다 새로 쓰이며(회전), 이 파일이 전달 수단이다.
type Endpoint struct {
	Bind  string `json:"bind"`
	Port  int    `json:"port"`
	Path  string `json:"path"`
	Token string `json:"token"`
}

// newToken은 32바이트 암호학 난수를 hex로 인코딩한 세션 토큰을 만든다.
func newToken() (string, error) {
	var b [32]byte
	if _, err := rand.Read(b[:]); err != nil {
		return "", fmt.Errorf("ide ipc token: %w", err)
	}
	return hex.EncodeToString(b[:]), nil
}

// endpointPath는 stateDir 아래 디스커버리 파일의 절대 경로를 반환한다.
func endpointPath(stateDir string) string {
	return filepath.Join(stateDir, EndpointFileName)
}

// writeEndpoint는 토큰이 포함된 endpoint를 0600 파일로 덮어쓴다.
// 실패하면 확장이 이전 토큰을 읽지 못하도록 기존 파일을 지운다.
func writeEndpoint(stateDir string, ep Endpoint) error {
	if stateDir == "" {
		return fmt.Errorf("ide ipc state dir is empty")
	}
	path := endpointPath(stateDir)
	raw, err := json.MarshalIndent(ep, "", "  ")
	if err != nil {
		return err
	}
	if err := os.WriteFile(path, raw, 0o600); err != nil {
		_ = os.Remove(path)
		return fmt.Errorf("write ide ipc endpoint: %w", err)
	}
	return nil
}

// removeEndpoint는 에이전트 종료 시 디스커버리 파일을 지워 죽은 포트로 붙지 않게 한다.
func removeEndpoint(stateDir string) {
	if stateDir == "" {
		return
	}
	_ = os.Remove(endpointPath(stateDir))
}

// ReadEndpoint는 stateDir의 ide-ipc.json을 읽는다.
// 파일이 없거나 JSON이 아니면 오류다. 호출자가 토큰을 로그에 남기면 안 된다.
func ReadEndpoint(stateDir string) (Endpoint, error) {
	var ep Endpoint
	raw, err := os.ReadFile(endpointPath(stateDir))
	if err != nil {
		return Endpoint{}, err
	}
	if err := json.Unmarshal(raw, &ep); err != nil {
		return Endpoint{}, fmt.Errorf("parse ide ipc endpoint: %w", err)
	}
	if ep.Bind == "" || ep.Port <= 0 || ep.Path == "" || ep.Token == "" {
		return Endpoint{}, fmt.Errorf("incomplete ide ipc endpoint")
	}
	return ep, nil
}
